### ISSUE TITLE
Support for arbitrary aspect ratios in output window

### ISSUE THREAD

**Issue by alphalambda** [google/codeworld#938]
I wonder how difficult it would be to have support for arbitrary aspect ratios in the output, because a square output limits student possibilities to create interesting scenes, since most devices are actually not square. Restricting the window to be square forces them to waste a lot of space for their scene in the horizontal direction, especially when compared with similar environments that target the same population.

One of the things I like the most about CodeWorld is the fact that the output uses abstract display units instead of pixels or other device-dependent units. Also, I think 10x10 is a good default choice for a general drawing. However, I don't see why choosing to build a 15x10 scene would be a problem, especially now that there are built-in controls to zoom and pan the output.

The only possible problem could be the `coordinatePlane` facility, which I think should be a function anyway, so that the span and resolution are configurable. This could be also important for using CodeWorld as a programmable graphing calculator, for which it has a lot of potential.

**Comment by cdsmith** [comment]
The reason for a fixed-size canvas has always been simplicity.  I agree with most of what you're saying, but would just need to see what the UI looks like.  Adding arguments to a function is I think a big step in complexity: `drawingOf(flower)` can be read and pronounced naturally, while `drawingOf(flower, 10, 10)` feels a lot more technical.

Maybe there's a clever way to do this without making the code complex.  Maybe even something in the UI, where you can select a device to build for outside of the code, and that includes the aspect ratio.

**Comment by alphalambda** [comment]
I don't think that `drawingOf` needs to be modified. On the CodeWorld language side, the only modification I think is needed is adding a `wideCoordinatePlane` that takes in a `width` as a parameter, so that `wideCoordinatePlane(width)` fills a rectangular area of `width` by 20 units.

Then, in the GUI side, maybe a small button to *unlock* the aspect, so that when you move the left side of the output window while it is unlocked, it changes only the width. 

Finally, for the fullscreen mode, maybe a parameter in the URL, like `... mode=codeworld&wide=30`.

I think that is all that is needed, as a drawing is potentially infinite, so it is just a convention that you create the interesting part of your drawing so that it fits the expected output.

**Comment by cdsmith** [comment]
Okay, this sounds fine.  Although I might prefer one coordinatePlane Picture value that is conceptually infinite (of course, the drawing code would look at the transformations and decide how much needs to be drawn).  This might be unfortunate for current projects that rely on the bounded coordinate plane, though.

**Comment by alphalambda** [comment]
The problem I see with an infinite `coordinatePlane` is that it would be a magic value (like the current `codeWorldLogo`) that is not possible to  create within CodeWorld, unless there are also new API functions to facilitate the creation of infinite Pictures. Otherwise, a user could not create their own version of `coordinatePlane`. The problem an infinite `coordinatePlane` would solve is easily solved currently in user space, as I illustrated with the `graphed` function.

**Comment by alphalambda** [comment]
I gave it more thought, and the problem I have is that we may be creating confusion similar to the Text/Lettering dichotomy. A coordinate plane (which is an infinite abstract concept) and a particular pictorial representation of a finite part of a coordinate plane are not the same thing, in the same vein that a Text and a Lettering of it are not the same thing. Just like there are many styled letterings of the same text, which may come in different colors, fonts and sizes, there could be many styled renderings of an abstract coordinate plane (with or without guides, labels, with different spacing, etc.).

**Comment by cdsmith** [comment]
There's plenty of time to discuss the coordinate plane problem in the other issue.  I actually think it would be just fine to implement this without getting to that question at all.  So after the implementation, `coordinatePlane` as a picture just might not take up the entire screen any longer.

A few design points:

1. I like the idea of having a "Lock aspect ratio" toggle, which controls the screen.  If you unlock the aspect ratio, resize, and then relock, I think the screen should remain fixed at the non-square ratio.
2. Would you want to define a unit of distance as 1/20 the smaller of the two screen dimensions, so that if the screen is non-square, you get extra distance in the longer dimension?
3. Ideally, when exporting a project, it should export with the aspect ratio properties preserved.  So, if the aspect ratio is unlocked, you'd get a link that resizes the canvas to the whole screen; if it's locked, it would resize the canvas to the largest that fits at that aspect ratio.

**Comment by alphalambda** [comment]
About point 2:
I was thinking about allowing just 1 degree of freedom instead of 2. So, the vertical dimension would stay at 20 units and the horizontal dimension would vary. I think that is simpler, and it covers all possible aspect ratios anyway. 

The good think about a fixed 20 by 20 display port is that it helps students focus better. A fully unconstrained view port could be more distracting.
